python/pyspark/sql/observation.py

```diff
@@ -0,0 +1,147 @@
+        """Constructs a named or unnamed Observation instance.
+
+        Parameters
+        ----------
+        name : str, optional
+            default is a random UUID string. This is the name of the Observation and the metric.
+        """
+        assert isinstance(name, basestring), "name should be a string"
+        self._name = name or str(uuid4())
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df, *exprs):
+        """Attaches this observation to the given :class:`DataFrame` to observe aggregations.
+
+        Parameters
+        ----------
+        df : :class:`DataFrame`
+            the :class:`DataFrame` to be observed
+        exprs : list of :class:`Column`
+            column expressions (:class:`Column`).
+
+        Returns
+        -------
+        :class:`DataFrame`
+            the observed :class:`DataFrame`.
+        """
+        assert exprs, "exprs should not be empty"
+        assert all(isinstance(c, Column) for c in exprs), "all exprs should be Column"
+        assert self._jo is None, "an Observation can be used with a DataFrame only once"
+
+        self._jvm = df._sc._jvm
+        self._jo = self._jvm.org.apache.spark.sql.Observation(self._name)
+        observed_df = self._jo.on(df._jdf,
+                                  exprs[0]._jc,
+                                  column._to_seq(df._sc, [c._jc for c in exprs[1:]]))
+        return DataFrame(observed_df, df.sql_ctx)
+
+    @property
+    def get(self):
+        """Get the observed metrics.
+
+        Waits until the observed dataset finishes its first action. Only the result of the
+        first action is available. Subsequent actions do not modify the result.
+
+        Returns
+        -------
+        :class:`Row`
+            the observed metrics
+        """
+        assert self._jo is not None, 'call DataFrame.observe'
+        jrow = self._jo.get()
+        return self._to_row(jrow)
+
+    def _to_row(self, jrow):
+        field_names = jrow.schema().fieldNames()
+        values_scala_map = jrow.getValuesMap(self._jvm.PythonUtils.toSeq(list(field_names)))
+        values_java_map = self._jvm.scala.collection.JavaConversions.mapAsJavaMap(values_scala_map)
+        return Row(**values_java_map)
+
+
+def _test():
+    import doctest
+    import sys
+    from pyspark.context import SparkContext
+    from pyspark.sql import SparkSession
+    from pyspark.sql.types import StructType, StructField, IntegerType, StringType
+    import pyspark.sql.observation
+    globs = pyspark.sql.observation.__dict__.copy()
+    sc = SparkContext('local[4]', 'PythonTest')
+    globs['sc'] = sc
+    globs['spark'] = SparkSession(sc)
+    globs['df'] = sc.parallelize([(2, 'Alice'), (5, 'Bob')]) \
+        .toDF(StructType([StructField('age', IntegerType()),
+                          StructField('name', StringType())]))
+
+    (failure_count, test_count) = doctest.testmod(pyspark.sql.observation, globs=globs)
+    globs['sc'].stop()
```

Comment:
```suggestion
    sc.stop()
```

Reply:
the self-contained example now needs `sc`, so I keep this